pkg/operator/ceph/object/topic/provisioner.go

```diff
@@ -0,0 +1,240 @@
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create session for bucket topic: %q provisioning", topic.Name)
+	}
+
+	attr := make(map[string]*string)
+
+	attr["push-endpoint"] = &topic.Spec.Endpoint
+	attr["OpaqueData"] = &topic.Spec.OpaqueData
+	persistent := strconv.FormatBool(topic.Spec.Persistent)
+	attr["persistent"] = &persistent
+	var verifySSL string
+	var useSSL string
+	if topic.Spec.AMQP != nil {
+		attr["amqp-exchange"] = &topic.Spec.AMQP.Exchange
+		attr["amqp-ack-level"] = &topic.Spec.AMQP.AckLevel
+		verifySSL = strconv.FormatBool(topic.Spec.AMQP.VerifySSL)
+		attr["verify-ssl"] = &verifySSL
+	}
+	if topic.Spec.HTTP != nil {
+		verifySSL = strconv.FormatBool(topic.Spec.HTTP.VerifySSL)
+		attr["verify-ssl"] = &verifySSL
+	}
+	if topic.Spec.Kafka != nil {
+		useSSL = strconv.FormatBool(topic.Spec.Kafka.UseSSL)
+		attr["use-ssl"] = &useSSL
+		attr["kafka-ack-level"] = &topic.Spec.Kafka.AckLevel
+		verifySSL = strconv.FormatBool(topic.Spec.Kafka.VerifySSL)
+		attr["verify-ssl"] = &verifySSL
+	}
+
+	topicOutput, err := createSNSClient(sess).CreateTopic(&sns.CreateTopicInput{
+		Name:       &topic.Name,
+		Attributes: attr,
+	})
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to provisioning bucket topic: %q", topic.Name)
+	}
+
+	logger.Infof("topic created with ARN: %q", *topicOutput.TopicArn)
+
+	return topicOutput.TopicArn, nil
+}
+
+func (p *Provisioner) Delete(topic *cephv1.CephBucketTopic) error {
+	logger.Infof("deleting topic: %q", topic.Name)
+	if topic.Status.ARN == nil {
+		logger.Warningf("ignore topic deletion. topic %q was never successfully provisioned", topic.Name)
+		return nil
+	}
+
+	sess, err := p.createSession(types.NamespacedName{Name: topic.Spec.ObjectStoreName, Namespace: topic.Spec.ObjectStoreNamespace})
+	if err != nil {
+		return errors.Wrapf(err, "failed to create session for bucket topic: %q deletion", topic.Name)
+	}
+
+	_, err = createSNSClient(sess).DeleteTopic(&sns.DeleteTopicInput{TopicArn: topic.Status.ARN})
+
+	if err != nil {
+		if err.(awserr.Error).Code() != sns.ErrCodeNotFoundException {
+			return errors.Wrapf(err, "failed to delete bucket topic: %q", topic.Name)
+		}
+		logger.Warningf("ignore topic deletion. topic %q was already deleted", topic.Name)
+	}
+
+	logger.Infof("topic %q deleted", topic.Name)
+
+	return nil
+}
+
+func GetARN(bucketTopic *cephv1.CephBucketTopic) (string, error) {
+	if bucketTopic.Status == nil || bucketTopic.Status.ARN == nil {
+		return "", errors.Errorf("no ARN in topic. CephBucketTopic: %q was not provisioned successfully", bucketTopic.Name)
+	}
+	topicARN := *bucketTopic.Status.ARN
+	parsedTopicARN, err := arn.Parse(topicARN)
+	if err != nil {
+		return topicARN, errors.Wrapf(err, "failed to parse topic ARN: %q", topicARN)
+	}
+	if strings.ToLower(parsedTopicARN.Service) != "sns" {
+		return topicARN, errors.Errorf("topic ARN: %q must have 'sns' service", topicARN)
```

Comment:
this can only happen if there is some RGW bug. don't think that the user can control that

Reply:
It may be more clear to have the message be: `failed to retrieve "sns" service for CephBucketTopic %q ARN %q, instead got [%v]`. Where %v is filled with `parsedTopicARN.Service`.